test_security/test/test_invalid_secure_node_creation_c.cpp

```diff
@@ -81,74 +101,87 @@ class CLASSNAME (TestInvalidSecureNode, RMW_IMPLEMENTATION) : public ::testing::
 
   void TearDown()
   {
-    rcl_ret_t ret = rcl_node_fini(this->node_ptr);
-    delete this->node_ptr;
-    ret = rcl_shutdown(&context);
-    (void)ret;
     custom_putenv("ROS_SECURITY_ROOT_DIRECTORY", previous_root_dir);
   }
+};
 
-  void test_node_creation(
-    const char * ROS_SECURITY_ROOT_DIRECTORY, const char * ROS_SECURITY_ENABLE,
-    const char * ROS_SECURITY_STRATEGY,
-    const char * context_name, bool should_fail_context_creation,
-    bool should_fail_node_creation)
-  {
-    if (ROS_SECURITY_ROOT_DIRECTORY != NULL) {
-      custom_putenv("ROS_SECURITY_ROOT_DIRECTORY", ROS_SECURITY_ROOT_DIRECTORY);
-    }
-    custom_putenv("ROS_SECURITY_ENABLE", ROS_SECURITY_ENABLE);
-    custom_putenv("ROS_SECURITY_STRATEGY", ROS_SECURITY_STRATEGY);
-    rcl_ret_t ret;
-    rcl_init_options_t init_options = rcl_get_zero_initialized_init_options();
-    ret = rcl_init_options_init(&init_options, rcl_get_default_allocator());
+using TestSecureNodes = CLASSNAME(TestSecureNodes, RMW_IMPLEMENTATION);
+
+TEST_P(TestSecureNodes, test_invalid_keystore) {
+  const auto & test_config = GetParam();
+  if (test_config.ROS_SECURITY_ROOT_DIRECTORY != NULL) {
+    custom_putenv("ROS_SECURITY_ROOT_DIRECTORY", test_config.ROS_SECURITY_ROOT_DIRECTORY);
+  }
+  custom_putenv("ROS_SECURITY_ENABLE", test_config.ROS_SECURITY_ENABLE);
+  custom_putenv("ROS_SECURITY_STRATEGY", test_config.ROS_SECURITY_STRATEGY);
+  rcl_ret_t ret;
+  rcl_init_options_t init_options = rcl_get_zero_initialized_init_options();
+  ret = rcl_init_options_init(&init_options, rcl_get_default_allocator());
+  ASSERT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+  context = rcl_get_zero_initialized_context();
+  const char * argv[] = {"--ros-args", "--security-context", test_config.context_name};
+  ret = rcl_init(sizeof(argv) / sizeof(char *), argv, &init_options, &context);
+  if (test_config.should_fail_context_creation) {
+    ASSERT_EQ(RCL_RET_ERROR, ret) << rcl_get_error_string().str;
+    return;
+  }
+  ASSERT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+  this->node = rcl_get_zero_initialized_node();
+  rcl_node_options_t node_options = rcl_node_get_default_options();
+  ret = rcl_node_init(
+    &this->node, "test_invalid_secure_node_creation_c", "", &context, &node_options);
+  if (test_config.should_fail_node_creation) {
+    ASSERT_EQ(RCL_RET_ERROR, ret) << rcl_get_error_string().str;
```

Comment:
@ivanpauno hmm if this assertion fails, the context will never be shutdown. Consider using OSRF testing tools' macros for cleanup on exit.

Reply:
Makes sense.

Reply:
See https://github.com/ros2/system_tests/pull/406/commits/5e707e37769a795f19ebf9b721a26ca57b962511.